nova_vm/src/ecmascript/builtins/indexed_collections/array_objects/array_prototype.rs

```diff
@@ -3247,13 +3247,80 @@ impl ArrayPrototype {
         Ok(ArrayIterator::from_object(agent, o, CollectionIteratorKind::Value).into_value())
     }
 
+    /// ### [23.1.3.39 Array.prototype.with ( index, value )](https://tc39.es/ecma262/#sec-array.prototype.with)
     fn with(
-        _agent: &mut Agent,
-        _gc: GcScope<'_, '_>,
-        _this_value: Value,
-        _: ArgumentsList,
+        agent: &mut Agent,
+        mut gc: GcScope<'_, '_>,
+        this_value: Value,
+        arguments: ArgumentsList,
     ) -> JsResult<Value> {
-        todo!();
+        // Fast path: Array is dense and contains no descriptors. No JS
+        // functions can thus be called by with.
+        if let Value::Array(array) = this_value {
+            let len = array.len(agent) as i64;
+            let index = arguments.get(0);
+            let relative_index =
+                to_integer_or_infinity(agent, gc.reborrow(), index)?.into_i64(agent);
+            let actual_index = if relative_index >= 0 {
+                relative_index
+            } else {
+                len + relative_index
+            };
+            if array.is_trivial(agent)
+                && array.is_dense(agent)
+                && !(actual_index >= len || actual_index < 0)
```

Comment:
**nitpick:** Not relevant after adding the throw path but: I'd remove the double-negation here by going `0 <= actual_index && actual_index < len`.

Reply:
Out of curiosity, is this something that rustc and llvm could optimize?

Reply:
Yeah, I'm sure the resulting code is equivalent either way. It's just a matter of readability: Double negatives tend to confuse people :)